### ISSUE TITLE
Strict aliasing violation warnings for simd-intrinsics.c in SSE2 build

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4446]
In an `--enable-simd=sse2` build made with `gcc version 4.4.7 20120313 (Red Hat 4.4.7-23)`:

```
simd-intrinsics.c: In function 'SIMDSHA1body':
simd-intrinsics.c:1368: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:1360: note: initialized from here
simd-intrinsics.c:1373: note: initialized from here
simd-intrinsics.c:1378: note: initialized from here
simd-intrinsics.c:1368: note: initialized from here
simd-intrinsics.c:1373: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:1360: note: initialized from here
simd-intrinsics.c:1368: note: initialized from here
simd-intrinsics.c:1378: note: initialized from here
simd-intrinsics.c:1373: note: initialized from here
simd-intrinsics.c:1378: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:1360: note: initialized from here
simd-intrinsics.c:1368: note: initialized from here
simd-intrinsics.c:1373: note: initialized from here
simd-intrinsics.c:1378: note: initialized from here
simd-intrinsics.c: In function 'SIMDmd4body':
simd-intrinsics.c:853: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:845: note: initialized from here
simd-intrinsics.c:858: note: initialized from here
simd-intrinsics.c:863: note: initialized from here
simd-intrinsics.c:853: note: initialized from here
simd-intrinsics.c:858: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:845: note: initialized from here
simd-intrinsics.c:853: note: initialized from here
simd-intrinsics.c:863: note: initialized from here
simd-intrinsics.c:858: note: initialized from here
simd-intrinsics.c:863: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:845: note: initialized from here
simd-intrinsics.c:853: note: initialized from here
simd-intrinsics.c:858: note: initialized from here
simd-intrinsics.c:863: note: initialized from here
simd-intrinsics.c: In function 'SIMDmd5body':
simd-intrinsics.c:168: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:160: note: initialized from here
simd-intrinsics.c:173: note: initialized from here
simd-intrinsics.c:178: note: initialized from here
simd-intrinsics.c:168: note: initialized from here
simd-intrinsics.c:173: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:160: note: initialized from here
simd-intrinsics.c:168: note: initialized from here
simd-intrinsics.c:178: note: initialized from here
simd-intrinsics.c:173: note: initialized from here
simd-intrinsics.c:178: warning: dereferencing pointer 'p' does break strict-aliasing rules
simd-intrinsics.c:160: note: initialized from here
simd-intrinsics.c:168: note: initialized from here
simd-intrinsics.c:173: note: initialized from here
simd-intrinsics.c:178: note: initialized from here
```

I haven't looked into these yet.

**Comment by magnumripper** [comment]
I'm puzzled we've never seen this warning before. I'm pretty sure I haven't, or I'd have fixed it.

Here's the first one (as in lowest line number). Rest of them most likely virtually identical:
```c
  void SIMDmd5body(vtype* _data, unsigned int *out, uint32_t *reload_state, unsigned SSEi_flags)
  {
          vtype w[16*SIMD_PARA_MD5];                                          /* Original declaration */
(...)
          if (SSEi_flags & SSEi_FLAT_IN) {
                  // Move _data to __data, mixing it SIMD_COEF_32 wise.
                  unsigned j, k;
                  uint32_t *p = (uint32_t*)w;                                  /* This is line 160 */
                  vtype *W = w;
                  uint32_t *saved_key = (uint32_t*)_data;
                  MD5_PARA_DO(k)
                  {
                          if (SSEi_flags & SSEi_4BUF_INPUT) {
                                  for (j=0; j < 16; j++)
                                          for (i=0; i < VS32; i++)
                                                  *p++ = saved_key[(i<<6)+j];  /* Line 168 (similar in 173, 178) */
                                  saved_key += (VS32<<6);
                          } else (...)
```
I guess we should just change the original declaration of `vtype w[16*SIMD_PARA_MD5];` into something like
```c
    union {
        vtype vec[16*SIMD_PARA_MD5];
        uint32_t u32[1];
    } w;
```
(IIRC we should be able to say `uint32_t u32[0]` but some compilers choke on it (at least in OpenCL) and it doesn't matter anyway)

...and then `s/\bw\b/w.vec/g;` accordingly. Then change declaration of p to `uint32_t *p = w.u32;`. Maybe it isn't silly although in this arch-specific intrinsics case I think it is. Silly or not, it should solve the problem with no ill effects.

Line 162 is not mentioned. If that one is (also?) a problem, we'd need to change the parameter to `void* _data`. The fact that the compiler seem to swap the messages "initialized from here" and "dereferencing pointer 'p'" around doesn't make me certain we're getting proper warnings at all - maybe `_data` vs. `saved_key` is also a problem or even *the* problem.

And conversely, if line 162 is not a problem, I fail to understand why line 160 would be one.

**Comment by magnumripper** [comment]
On a side note, `unsigned j, k` would be nicer written `unsigned int j, k` but it wouldn't change the result

**Comment by magnumripper** [comment]
https://stackoverflow.com/a/46752535 is a good read (even if not applicable here)
> Using a __m128i* to load from an array of int is safe because the __m128 types are defined to allow aliasing just like ISO C unsigned char*. (e.g. in gcc's headers, the definition includes `__attribute__((may_alias))`.)
>
> The reverse isn't safe (pointing an int* onto part of a __m128i object). MSVC guarantees that's safe, but GCC/clang don't. (-fstrict-aliasing is on by default). It sometimes works with GCC/clang, but why risk it? It sometimes even interferes with optimization; see [this Q&A](https://stackoverflow.com/questions/63360909/avx2-simd-performs-relatively-worse-to-scalar-at-higher-optimization-level). See also [Is 'reinterpret_cast'ing between hardware SIMD vector pointer and the corresponding type an undefined behavior?](https://stackoverflow.com/questions/52112605/is-reinterpret-casting-between-hardware-simd-vector-pointer-and-the-correspond)

**Comment by magnumripper** [comment]
So we can simply do this:
```diff
diff --git a/src/simd-intrinsics.c b/src/simd-intrinsics.c
index 6a55f1a07..7078173f8 100644
--- a/src/simd-intrinsics.c
+++ b/src/simd-intrinsics.c
@@ -115,7 +115,7 @@ void md5_unreverse(uint32_t *hash)
 
 #undef INIT_A
 
-void SIMDmd5body(vtype* _data, unsigned int *out,
+void SIMDmd5body(uint32_t* _data, unsigned int *out,
                 uint32_t *reload_state, unsigned SSEi_flags)
 {
        vtype w[16*SIMD_PARA_MD5];
@@ -140,7 +140,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
 #if __SSE4_1__ || __MIC__
                unsigned k;
                vtype *W = w;
-               uint32_t *saved_key = (uint32_t*)_data;
+               uint32_t *saved_key = _data;
                MD5_PARA_DO(k)
                {
                        if (SSEi_flags & SSEi_4BUF_INPUT) {
@@ -159,7 +159,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                unsigned j, k;
                uint32_t *p = (uint32_t*)w;
                vtype *W = w;
-               uint32_t *saved_key = (uint32_t*)_data;
+               uint32_t *saved_key = _data;
                MD5_PARA_DO(k)
                {
                        if (SSEi_flags & SSEi_4BUF_INPUT) {
@@ -194,7 +194,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                // now set our data pointer to point to this 'mixed' data.
                data = w;
        } else
-               data = _data;
+               data = (vtype*)_data;
 
        if (!(SSEi_flags & SSEi_RELOAD))
        {
@@ -584,7 +584,7 @@ static MAYBE_INLINE void dispatch(unsigned char buffers[8][64*MD5_SSE_NUM_KEYS],
                                mmxput3(buffers, bufferid, length, 2, saltlen, f);
                                break;
                }
-               SIMDmd5body((vtype*)&buffers[bufferid], f, NULL, SSEi_MIXED_IN);
+               SIMDmd5body((uint32_t*)&buffers[bufferid], f, NULL, SSEi_MIXED_IN);
                if (j++ < 1000 % 42 - 1)
                        continue;
                if (j == 1000 % 42) {
```
I tried this, no warning seen from gcc 7.5.0 (no warnings before it either) and after stripping I have an **identical** `simd-intrinsics.o`. I'll produce a complete patch.

**EDIT** that didn't help with gcc 4.4.7 at all

**Comment by magnumripper** [comment]
This does help with gcc 4.4.7
```diff
diff --git a/src/simd-intrinsics.c b/src/simd-intrinsics.c
index 6a55f1a..3c6fa60 100644
--- a/src/simd-intrinsics.c
+++ b/src/simd-intrinsics.c
@@ -118,7 +118,10 @@ void md5_unreverse(uint32_t *hash)
 void SIMDmd5body(vtype* _data, unsigned int *out,
                 uint32_t *reload_state, unsigned SSEi_flags)
 {
-       vtype w[16*SIMD_PARA_MD5];
+       union {
+               vtype vec[16*SIMD_PARA_MD5];
+               uint32_t u32[1];
+       } w;
        vtype a[SIMD_PARA_MD5];
        vtype b[SIMD_PARA_MD5];
        vtype c[SIMD_PARA_MD5];
@@ -139,7 +142,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                // Move _data to __data, mixing it SIMD_COEF_32 wise.
 #if __SSE4_1__ || __MIC__
                unsigned k;
-               vtype *W = w;
+               vtype *W = w.vec;
                uint32_t *saved_key = (uint32_t*)_data;
                MD5_PARA_DO(k)
                {
@@ -157,8 +160,8 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                }
 #else
                unsigned j, k;
-               uint32_t *p = (uint32_t*)w;
-               vtype *W = w;
+               uint32_t *p = w.u32;
+               vtype *W = w.vec;
                uint32_t *saved_key = (uint32_t*)_data;
                MD5_PARA_DO(k)
                {
@@ -192,7 +195,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                }
 #endif
                // now set our data pointer to point to this 'mixed' data.
-               data = w;
+               data = w.vec;
        } else
                data = _data;
 
```
(Similar changes needs to be applied for MD4 and SHA-1)

**Comment by magnumripper** [comment]
Here's a simpler version
```diff
@@ -118,7 +118,11 @@ void md5_unreverse(uint32_t *hash)
 void SIMDmd5body(vtype* _data, unsigned int *out,
                 uint32_t *reload_state, unsigned SSEi_flags)
 {
-       vtype w[16*SIMD_PARA_MD5];
+       union {
+               vtype vec[16*SIMD_PARA_MD5];
+               uint32_t u32[1];
+       } uw;
+       vtype *w = uw.vec;
        vtype a[SIMD_PARA_MD5];
        vtype b[SIMD_PARA_MD5];
        vtype c[SIMD_PARA_MD5];
@@ -157,7 +161,7 @@ void SIMDmd5body(vtype* _data, unsigned int *out,
                }
 #else
                unsigned j, k;
-               uint32_t *p = (uint32_t*)w;
+               uint32_t *p = uw.u32;
                vtype *W = w;
                uint32_t *saved_key = (uint32_t*)_data;
                MD5_PARA_DO(k)
```
~For some reason this (when applied to MD4 and SHA-1 as well) results in somewhat different code for SHA-1 on my gear (using gcc-10). It can't be seen using the raw format but benchmarking pbkdf2-hmac-sha1 it appears I got a 5% loss in performance. MD4 and MD5 ends up with unchanged code. For what it's worth `objdump -d` produces an additional 80 lines of output.~ 

**EDIT: That last thing was due to a flaw in my test conditions**

**Comment by magnumripper** [comment]
Here's a longer diff for SHA-1 that doesn't end up in different code on my gear but still fixes the problem for gcc-4.4.7:
```diff
@@ -1141,31 +1147,31 @@ void SIMDmd4body(vtype* _data, unsigned int *out, uint32_t *reload_state,
     tmp[i] = vternarylogic(data[i*16+t-3], data[i*16+t-8],  \
                            data[i*16+t-14], 0x96);          \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );               \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2b(t)                                        \
-    tmp[i] = vternarylogic(w[i*16+((t-3)&0xF)], data[i*16+t-8], \
+    tmp[i] = vternarylogic(w.vec[i*16+((t-3)&0xF)], data[i*16+t-8], \
                            data[i*16+t-14], 0x96);              \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );                   \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2c(t)                                                \
-    tmp[i] = vternarylogic(w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)],    \
+    tmp[i] = vternarylogic(w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)],    \
                            data[i*16+t-14], 0x96);                      \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );                           \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2d(t)                                                \
-    tmp[i] = vternarylogic(w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)],    \
-                           w[i*16+((t-14)&0xF)], 0x96);                 \
+    tmp[i] = vternarylogic(w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)],    \
+                           w.vec[i*16+((t-14)&0xF)], 0x96);                 \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );                           \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2(t)                                                 \
-    tmp[i] = vternarylogic(w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)],    \
-                           w[i*16+((t-14)&0xF)], 0x96);                 \
-    tmp[i] = vxor( tmp[i], w[i*16+((t-16)&0xF)] );                      \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    tmp[i] = vternarylogic(w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)],    \
+                           w.vec[i*16+((t-14)&0xF)], 0x96);                 \
+    tmp[i] = vxor( tmp[i], w.vec[i*16+((t-16)&0xF)] );                      \
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #else
 
@@ -1173,31 +1179,31 @@ void SIMDmd4body(vtype* _data, unsigned int *out, uint32_t *reload_state,
     tmp[i] = vxor( data[i*16+t-3], data[i*16+t-8] );    \
     tmp[i] = vxor( tmp[i], data[i*16+t-14] );           \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );           \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2b(t)                                    \
-    tmp[i] = vxor( w[i*16+((t-3)&0xF)], data[i*16+t-8] );   \
+    tmp[i] = vxor( w.vec[i*16+((t-3)&0xF)], data[i*16+t-8] );   \
     tmp[i] = vxor( tmp[i], data[i*16+t-14] );               \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );               \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2c(t)                                        \
-    tmp[i] = vxor( w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)] );  \
+    tmp[i] = vxor( w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)] );  \
     tmp[i] = vxor( tmp[i], data[i*16+t-14] );                   \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );                   \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2d(t)                                        \
-    tmp[i] = vxor( w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)] );  \
-    tmp[i] = vxor( tmp[i], w[i*16+((t-14)&0xF)] );              \
+    tmp[i] = vxor( w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)] );  \
+    tmp[i] = vxor( tmp[i], w.vec[i*16+((t-14)&0xF)] );              \
     tmp[i] = vxor( tmp[i], data[i*16+t-16] );                   \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 
 #define SHA1_EXPAND2(t)                                         \
-    tmp[i] = vxor( w[i*16+((t-3)&0xF)], w[i*16+((t-8)&0xF)] );  \
-    tmp[i] = vxor( tmp[i], w[i*16+((t-14)&0xF)] );              \
-    tmp[i] = vxor( tmp[i], w[i*16+((t-16)&0xF)] );              \
-    w[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
+    tmp[i] = vxor( w.vec[i*16+((t-3)&0xF)], w.vec[i*16+((t-8)&0xF)] );  \
+    tmp[i] = vxor( tmp[i], w.vec[i*16+((t-14)&0xF)] );              \
+    tmp[i] = vxor( tmp[i], w.vec[i*16+((t-16)&0xF)] );              \
+    w.vec[i*16+((t)&0xF)] = vroti_epi32(tmp[i], 1);
 #endif
 
 #define SHA1_ROUND2a(a,b,c,d,e,F,t)                 \
@@ -1255,7 +1261,7 @@ void SIMDmd4body(vtype* _data, unsigned int *out, uint32_t *reload_state,
         tmp[i] = vroti_epi32(a[i], 5);              \
         e[i] = vadd_epi32( e[i], tmp[i] );          \
         e[i] = vadd_epi32( e[i], cst );             \
-        e[i] = vadd_epi32( e[i], w[i*16+(t&0xF)] ); \
+        e[i] = vadd_epi32( e[i], w.vec[i*16+(t&0xF)] ); \
         b[i] = vroti_epi32(b[i], 30);               \
         SHA1_EXPAND2(t+16)                          \
     }
@@ -1267,7 +1273,7 @@ void SIMDmd4body(vtype* _data, unsigned int *out, uint32_t *reload_state,
         tmp[i] = vroti_epi32(a[i], 5);              \
         e[i] = vadd_epi32( e[i], tmp[i] );          \
         e[i] = vadd_epi32( e[i], cst );             \
-        e[i] = vadd_epi32( e[i], w[i*16+(t&0xF)] ); \
+        e[i] = vadd_epi32( e[i], w.vec[i*16+(t&0xF)] ); \
         b[i] = vroti_epi32(b[i], 30);               \
     }
 #define INIT_D 0x10325476
@@ -1303,7 +1309,10 @@ void sha1_unreverse3(uint32_t *hash)
 void SIMDSHA1body(vtype* _data, uint32_t *out, uint32_t *reload_state,
                  unsigned SSEi_flags)
 {
-	vtype w[16*SIMD_PARA_SHA1];
+	union {
+		vtype vec[16*SIMD_PARA_SHA1];
+		uint32_t u32[1];
+	} w;
 	vtype a[SIMD_PARA_SHA1];
 	vtype b[SIMD_PARA_SHA1];
 	vtype c[SIMD_PARA_SHA1];
@@ -1318,7 +1327,7 @@ void SIMDSHA1body(vtype* _data, uint32_t *out, uint32_t *reload_state,
 		// Move _data to __data, mixing it SIMD_COEF_32 wise.
 #if __SSE4_1__ || __MIC__
 		unsigned k;
-		vtype *W = w;
+		vtype *W = w.vec;
 		uint32_t *saved_key = (uint32_t*)_data;
 		SHA1_PARA_DO(k)
 		{
@@ -1357,8 +1366,8 @@ void SIMDSHA1body(vtype* _data, uint32_t *out, uint32_t *reload_state,
 		}
 #else
 		unsigned j, k;
-		uint32_t *p = (uint32_t*)w;
-		vtype *W = w;
+		uint32_t *p = w.u32;
+		vtype *W = w.vec;
 		uint32_t *saved_key = (uint32_t*)_data;
 		SHA1_PARA_DO(k)
 		{
@@ -1393,7 +1402,7 @@ void SIMDSHA1body(vtype* _data, uint32_t *out, uint32_t *reload_state,
 #endif
 
 		// now set our data pointer to point to this 'mixed' data.
-		data = w;
+		data = w.vec;
 	} else
 		data = _data;
 
```
I really can't see why this happens.
